File: sdk-go/tests/bdd/test_world.go
package bdd

import (
	"os"

	decibel "github.com/gregnazario/decibel-sdks/sdk-go"
	"github.com/gregnazario/decibel-sdks/sdk-go/models"
)

// TestWorld maintains state across BDD scenario steps.
type TestWorld struct {
	// Configuration
	Config *decibel.DecibelConfig
	APIKey string

	// Clients
	ReadClient  *decibel.DecibelReadClient
	WriteClient *decibel.DecibelWriteClient

	// Error state
	LastError error

	// Market data
	Markets        []models.PerpMarketConfig
	MarketDepth    *models.MarketDepth
	MarketPrices   []models.MarketPrice
	Candlesticks   []models.Candlestick
	MarketContexts []models.MarketContext
	MarketTrades   []models.MarketTrade

	// Account data
	AccountOverview *models.AccountOverview
	Positions       []models.UserPosition
	OpenOrders      []models.UserOpenOrder
	OrderHistory    []models.UserOrderHistoryItem
	TradeHistory    []models.UserTradeHistoryItem
	FundingHistory  []models.UserFundingHistoryItem
	FundHistory     []models.UserFundHistoryItem
	Subaccounts     []models.UserSubaccount
	Delegations     []models.Delegation

	// TWAP data
	ActiveTwaps   []models.UserActiveTwap
	TwapHistory   []models.UserTwapHistoryItem

	// Vault data
	Vaults            []models.Vault
	UserVaults        []models.Vault
	VaultPerformances []models.VaultPerformance

	// Leaderboard
	Leaderboard []models.LeaderboardEntry

	// Test data
	TestMarketName     string
	TestSubaccountAddr string
}

// NewTestWorld creates a new TestWorld instance.
func NewTestWorld() *TestWorld {
	return &TestWorld{
		APIKey: os.Getenv("DECIBEL_API_KEY"),
	}
}

// Clear resets the test world state.
func (w *TestWorld) Clear() {
	w.Config = nil
	w.ReadClient = nil
	w.WriteClient = nil
	w.LastError = nil
	w.Markets = nil
	w.MarketDepth = nil
	w.MarketPrices = nil
	w.Candlesticks = nil
	w.MarketContexts = nil
	w.MarketTrades = nil
	w.AccountOverview = nil
	w.Positions = nil
	w.OpenOrders = nil
	w.OrderHistory = nil
	w.TradeHistory = nil
	w.FundingHistory = nil
	w.FundHistory = nil
	w.Subaccounts = nil
	w.Delegations = nil
	w.ActiveTwaps = nil
	w.TwapHistory = nil
	w.Vaults = nil
	w.UserVaults = nil
	w.VaultPerformances = nil
	w.Leaderboard = nil
	w.TestMarketName = ""
	w.TestSubaccountAddr = ""
}

// HasError returns true if there was an error.
func (w *TestWorld) HasError() bool {
	return w.LastError != nil
}

// SetError sets the last error.
func (w *TestWorld) SetError(err error) {
	w.LastError = err
}

// GetReadClient returns the read client, initializing if necessary.
func (w *TestWorld) GetReadClient() (*decibel.DecibelReadClient, error) {
	if w.ReadClient == nil {
		if w.Config == nil {
			w.Config = decibel.TestnetConfig()
		}
		client, err := decibel.NewDecibelReadClient(w.Config, w.APIKey)
		if err != nil {
			w.SetError(err)
			return nil, err
		}
		w.ReadClient = client
	}
	return w.ReadClient, nil
}

File: sdk-kotlin/src/test/kotlin/trade/decibel/sdk/bdd/TestWorld.kt
package trade.decibel.sdk.bdd

import trade.decibel.sdk.client.DecibelReadClient
import trade.decibel.sdk.config.DecibelConfig
import trade.decibel.sdk.models.*

/**
 * TestWorld maintains state across BDD scenario steps.
 */
class TestWorld {
    // Configuration
    var config: DecibelConfig? = null
    var apiKey: String? = System.getenv("DECIBEL_API_KEY")

    // Clients
    var readClient: DecibelReadClient? = null

    // Error state
    var lastError: Throwable? = null

    // Market data
    var markets: List<PerpMarketConfig>? = null
    var marketDepth: MarketDepth? = null
    var marketPrices: List<MarketPrice>? = null
    var candlesticks: List<Candlestick>? = null
    var marketContexts: List<MarketContext>? = null
    var marketTrades: List<MarketTrade>? = null

    // Account data
    var accountOverview: AccountOverview? = null
    var positions: List<UserPosition>? = null
    var openOrders: List<UserOpenOrder>? = null
    var orderHistory: List<UserOrderHistoryItem>? = null
    var tradeHistory: List<UserTradeHistoryItem>? = null
    var fundingHistory: List<UserFundingHistoryItem>? = null
    var fundHistory: List<UserFundHistoryItem>? = null
    var subaccounts: List<UserSubaccount>? = null
    var delegations: List<Delegation>? = null

    // TWAP data
    var activeTwaps: List<UserActiveTwap>? = null
    var twapHistory: List<UserTwapHistoryItem>? = null

    // Vault data
    var vaults: List<Vault>? = null
    var userVaults: List<Vault>? = null
    var vaultPerformances: List<VaultPerformance>? = null

    // Leaderboard
    var leaderboard: List<LeaderboardEntry>? = null

    // Test data
    var testMarketName: String = ""
    var testSubaccountAddr: String = ""

    /**
     * Resets the test world state.
     */
    fun clear() {
        config = null
        readClient = null
        lastError = null
        markets = null
        marketDepth = null
        marketPrices = null
        candlesticks = null
        marketContexts = null
        marketTrades = null
        accountOverview = null
        positions = null
        openOrders = null
        orderHistory = null
        tradeHistory = null
        fundingHistory = null
        fundHistory = null
        subaccounts = null
        delegations = null
        activeTwaps = null
        twapHistory = null
        vaults = null
        userVaults = null
        vaultPerformances = null
        leaderboard = null
        testMarketName = ""
        testSubaccountAddr = ""
    }

    /**
     * Returns true if there was an error.
     */
    fun hasError(): Boolean = lastError != null

    /**
     * Returns the read client, initializing if necessary.
     */
    fun getOrCreateReadClient(): DecibelReadClient {
        if (readClient == null) {
            if (config == null) {
                config = DecibelConfig.TESTNET
            }
            readClient = DecibelReadClient(config!!, apiKey)
        }
        return readClient!!
    }
}
